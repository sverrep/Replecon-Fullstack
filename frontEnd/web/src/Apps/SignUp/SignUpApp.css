
  .signup-container{
    display: flex;
    justify-content: center;
    width: 50%;
  }
  
  .vertical-center {
    margin: 0;
    top: 50%;
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
  }
  
  .signup-h3{
    text-align: center;
  }
  
  .signup-btns{
    margin-right: 5px;
    margin-left: 5px;
    margin-bottom: 15px;
    width: 60%;
  }
  
  .signup-btns-row{
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
  }

  .error-message{
      text-align: center;
      color:red;
  }