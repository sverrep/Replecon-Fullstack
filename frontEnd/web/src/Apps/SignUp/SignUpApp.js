import React from 'react';
import { withRouter, Redirect } from "react-router-dom";
import axios from 'axios'
import getIP from '../../settings.js';
import './SignUpApp.css';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Form from 'react-bootstrap/Form'
import FloatingLabel from 'react-bootstrap/FloatingLabel'
import {passwords} from '../../Components/passwords/mostCommon.js'

class SignUpApp extends React.Component {
    constructor(props) {
        super(props);
        this.state = { 
            email: '', 
            password: '', 
            name: '', 
            class_code: '', 
            first_name: '', 
            last_name: '', 
            redirect_login: false, 
            redirect_profile: false, 
            showError: false, 
            error: '',
            token: '',
        }
        this.handleChange = this.handleChange.bind(this)
        this.handleSignup = this.handleSignup.bind(this)
        this.handleReturnRedirect = this.handleReturnRedirect.bind(this)
        this.checkClassCode = this.checkClassCode.bind(this)
        this.validateStrongPassword = this.validateStrongPassword.bind(this)
        this.vaildateComplexity = this.vaildateComplexity.bind(this)
    }

    render() {
        if(this.state.redirect_login){
            return(
                <Redirect to={{pathname: '/Login'}}></Redirect>
            );
        }
        else if(this.state.redirect_profile){
            if(this.props.location.state.role === "Student")
            {
                return(
                    <Redirect to={{
                        pathname: '/Profile', 
                        state: { email: this.state.email, name: this.state.name, class_code: this.state.class_code, role: this.props.location.state.role, token: this.state.token }}}>
                    </Redirect>
                );
            }
            else if(this.props.location.state.role === "Teacher")
            {
                return(
                    <Redirect to={{
                        pathname: '/Profile', 
                        state: { role: this.props.location.state.role, token: this.state.token }}}>
                    </Redirect>
                );
            }
            
        }
        else{
            return (
                <Container className="signup-container">
                    <div className="vertical-center">
                        <Row>
                            <h3 className="signup-h3">{this.props.location.state.role} Sign Up</h3>
                        </Row>
                        <Row>
                            {this.setUpCommonFields()}
                            {this.setUpUniqueFields()}
                        </Row>
                        <p className="error-message">{this.state.error}</p>
                        <div className="signup-btns-row">
                            <Button className="signup-btns" onClick={() => this.handleSignup()}>
                            SignUp
                            </Button>
                        </div>
                        <div className="signup-btns-row">
                            <Button className="signup-btns" onClick={this.handleReturnRedirect}>
                            Already have an account? Log In
                            </Button>
                        </div>
                    </div>
                </Container>
        
            );
        }
    }

    setUpCommonFields(){
        return(
            <Row>
                <Form.Group className="mb-3" controlId="email">
                    <FloatingLabel label="Email">
                        <Form.Control id="email" placeholder="Email" onChange={this.handleChange} />
                    </FloatingLabel>
                    <FloatingLabel label="Password">
                        <Form.Control id="password" type="password" placeholder="Password" onChange={this.handleChange} />
                    </FloatingLabel>
                </Form.Group>
            </Row>
        )
    }
    setUpUniqueFields(){
        if(this.props.location.state.role === "Student"){
            return(
                <Row>
                    <Form.Group className="mb-3" controlId="email">
                        <FloatingLabel label="Name">
                            <Form.Control id="name" placeholder="Name" onChange={this.handleChange} />
                        </FloatingLabel>
                        <FloatingLabel label="Class Code">
                            <Form.Control id="class_code" placeholder="Class code" onChange={this.handleChange} />
                        </FloatingLabel>
                    </Form.Group>
                </Row>
            );
        }
        else if(this.props.location.state.role === "Teacher"){
            return( 
                <Row>
                    <Form.Group className="mb-3" controlId="email">
                        <FloatingLabel label="First Name">
                            <Form.Control id="first_name" placeholder="First Name" onChange={this.handleChange} />
                        </FloatingLabel>
                        <FloatingLabel label="Last Name">
                            <Form.Control id="last_name" placeholder="Last Name" onChange={this.handleChange} />
                        </FloatingLabel>
                    </Form.Group>
                </Row>
            );
        }
    }

    validateStrongPassword(password){
        var flag = true
        for(let i= 0; i <= 999; i++){
            if (password === passwords[i]){
                flag = false
            }
        }
        return flag
    }

    vaildateComplexity(password){
        let strongPassword = new RegExp('(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{8,})')
        if(strongPassword.test(password)){
            return true
        }
        else{
            return false
        }
    }

    async validateData() 
    {
        var reg = /^\w+([-]?\w+)*@\w+([-]?\w+)*(\.\w\w+)+$/;
        if (reg.test(this.state.email))
        {
            if(this.state.password.length === 0){
                this.setState({error: "Password needs to be longer", showError: true})
            }
            else{
                if(this.validateStrongPassword(this.state.password)===false){
                    this.setState({error: "Please use a stronger password", showError: true})
                }
                else{
                    
                    if(this.vaildateComplexity(this.state.password)===false){
                        this.setState({error: "Please use a more complex password. Must be 8 characters long and include at least 1 symbol, uppercase, and number", showError: true})
                    }
                    else{

                        if (this.props.location.state.role === "Student")
                        {
                            if(this.state.name === '')
                            {
                                this.setState({error: "Please enter a name"})
                                return false
                            }
                            else
                            {
                                if(this.state.class_code.length === 6)
                                {
                                    var check = await this.checkClassCode(this.state.class_code)
                                    if(check)
                                    {
                                        return true
                                    }
                                    else
                                    {
                                        this.setState({error: "Please enter an existing class code"})
                                        return false
                                    }        
                                }
                                else
                                {
                                    this.setState({error: "Class Code needs to be 6 characters long", showError: true})
                                    return false
                                }
                            }
                            
                        }
                        else if (this.props.location.state.role === "Teacher")
                        {
                            if(this.state.first_name === '')
                            {
                                this.setState({error: "Please enter a first name"})
                                return false
                            }
                            else
                            {
                                if(this.state.last_name === '')
                                {
                                    this.setState({error: "Please enter a last name"})
                                    return false
                                }
                                else
                                {
                                    return true
                                }
                            }
                        }
                    }
                }
            }
        }
        else
        {
            this.setState({error: "Email is invalid", showError: true})
            return false
        } 
    }

    async checkClassCode(class_code){
        var found = false
        await axios.get(getIP()+'/classrooms/')
        .then(response => {
            var classrooms = response.data
            for (let i = 0; i <= Object.keys(classrooms).length-1; i++)
            {
                if (classrooms[i].class_code === class_code)
                {
                    found = true
                }
            }
        })
        .catch(error => console.log(error))
        return found
    }
    
    handleChange(e) {
        const field = e.target.id
        this.setState({ [field] : e.target.value });
      }
    
    async handleSignup() {
        if(this.props.location.state.role === "Student")
        {
            if (await this.validateData()) {
                const payload = { username: this.state.email, password: this.state.password, first_name: this.state.name } 
                axios.post(getIP()+'/auth/register/', payload)
                .then(response => {
                    this.setState({ token: response.data.token })
                    axios.defaults.headers.common.Authorization = `Token ${this.state.token}`;
                    axios.post(getIP()+'/students/create/', {class_code: this.state.class_code})
                    .then(response => {
                        this.setState({ redirect_profile : true})
                    })
                    .catch(error => console.log(error));
                })            
                return
            }
            else {
                return
            }
        }
        else if(this.props.location.state.role === "Teacher"){
            if (await this.validateData()) {
                const payload = { username: this.state.email, password: this.state.password, first_name: this.state.first_name } 
                axios.post(getIP()+'/auth/register/', payload)
                .then(response => {
                    this.setState({ token: response.data.token })
                    axios.defaults.headers.common.Authorization = `Token ${this.state.token}`;
                    axios.post(getIP()+'/teachers/create/', {last_name: this.state.last_name})
                    .then(response => {
                        this.setState({ redirect_profile : true})
                    })
                    .catch(error => console.log(error));
                })            
                return
            }
            else {
                return
            }
        }
    }
    
    handleReturnRedirect(){
        this.setState({ redirect_login : true })
      }
  }

export default withRouter(SignUpApp)