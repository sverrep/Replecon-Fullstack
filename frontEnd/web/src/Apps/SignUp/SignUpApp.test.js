import React from 'react';
import SignUpApp from './SignUpApp.js'
import axios from 'axios'

jest.mock('axios');

const props = {
    location:{
        state:{
            role: "Student"
        }
    }  
}

const props2 = {
    location:{
        state:{
            role: "Teacher"
        }
    }  
}

describe('Sign Up Tests', ()=>{
    
    describe("Render Tests", ()=>{
        it('Teacher Render', ()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props2} />)
            expect(wrapper.getElements()).toMatchSnapshot();
        })

        it('Student Render', ()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            expect(wrapper.getElements()).toMatchSnapshot();
        })
    })

    describe('User Fields Test', ()=>{
        it('Teacher Unique Field ', ()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props2} />)
            const result = wrapper.instance().setUpUniqueFields()
            expect(result).toMatchSnapshot()
        })

        it('Student Unique Field ', ()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapper.instance().setUpUniqueFields()
            expect(result).toMatchSnapshot()
        })

        it('Common Field ', ()=>{
            let wrapperstudent = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapperstudent.instance().setUpCommonFields()

            let wrapperteacher = shallow(<SignUpApp.WrappedComponent {...props2} />)
            const result2 = wrapperteacher.instance().setUpCommonFields()
            expect(result2).toMatchSnapshot()
        })
    })

    describe('Password Tests', ()=>{
        it('should fail password complexity', ()=>{
            
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapper.instance().vaildateComplexity('12345')
            expect(result).toBe(false)
        });

        it('should validate password complexity', ()=>{

            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapper.instance().vaildateComplexity('!Student12')
            expect(result).toBe(true)
        });

        it('should validate strong password ', ()=>{

            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapper.instance().validateStrongPassword('todaywasagoodday142')
            expect(result).toBe(true)
        });

        it('should fail strong password ', ()=>{

            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            const result = wrapper.instance().validateStrongPassword('1234')
            expect(result).toBe(false)
        });
    })

    describe('User Validation', ()=>{
        it('should fail email validation', ()=>{
    
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            wrapper.instance().state.email = "student"
            wrapper.instance().validateData()
            expect(wrapper.instance().state.error).toBe("Email is invalid")
        })

        it('should fail class code length validation', ()=>{
    
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            wrapper.instance().state.email = "student@gmail.com"
            wrapper.instance().state.password = "!sdfsdfA12"
            wrapper.instance().state.name = "name"
            wrapper.instance().state.class_code = "12345"
            wrapper.instance().validateData()
            expect(wrapper.instance().state.error).toBe("Class Code needs to be 6 characters long")
        })

        it('should pass complete user validation', async()=>{
    
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            wrapper.instance().state.email = "student@gmail.com"
            wrapper.instance().state.password = "!sdfsdfA12"
            wrapper.instance().state.name = "name"
            wrapper.instance().state.class_code = "123456"
            wrapper.instance().checkClassCode = jest.fn()
            wrapper.instance().checkClassCode.mockReturnValueOnce(true)
            await wrapper.instance().validateData()
            expect(wrapper.instance().state.error).toBe("")
            
        })
        
    })

    describe('Class code API Calls', ()=>{
        it('should find class room', async()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)

            const classes = mockResponseData([
                {
                  "id":"1",
                  "class_name":"started",
                  "teacher_id":"Milestone",
                  "class_code":'123456'
                },
                {
                    "id":"1",
                    "class_name":"started",
                    "teacher_id":"Milestone",
                    "class_code":'1ASDD6'
                }])
                await axios.get.mockImplementationOnce(() =>  Promise.resolve(classes))
                const result = await wrapper.instance().checkClassCode('123456')
                expect(axios.get).toHaveBeenCalledTimes(1)
                expect(result).toBe(true)
                expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//classrooms/')
            
        })

        it('should not find classroom', async()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)

            const classes = mockResponseData([
                {
                  "id":"1",
                  "class_name":"started",
                  "teacher_id":"Milestone",
                  "class_code":'123457'
                },
                {
                    "id":"1",
                    "class_name":"started",
                    "teacher_id":"Milestone",
                    "class_code":'1ASDD6'
                }])
                await axios.get.mockImplementationOnce(() =>  Promise.resolve(classes))
                const result = await wrapper.instance().checkClassCode('123456')
                expect(result).toBe(false)
                expect(axios.get).toHaveBeenCalledTimes(1)
                expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//classrooms/')
        })
    })
    
    describe('SignUp API Calls', ()=>{
        it('should create student account and redirect to profile page', async()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props} />)
            wrapper.instance().validateData = jest.fn()
            wrapper.instance().validateData.mockReturnValueOnce(true)
            wrapper.instance().state.email = 'student@gmail.com'
            wrapper.instance().state.password = 'password'
            wrapper.instance().state.name = 'name'
            wrapper.instance().state.class_code = '123456'

            const payload = {
                username: 'student@gmail.com', 
                password: 'password', 
                first_name: 'name', 
                email: 'student@gmail.com'
            }

            const payload2 = {
                class_code:'123456'
            }

            const user = mockResponseData('Token')
            const user2 = mockResponseData(true)
            
            axios.post.mockImplementationOnce(() => Promise.resolve(user))
            axios.post.mockImplementationOnce(() => Promise.resolve(user2))

            await wrapper.instance().handleSignup()
            
            expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//auth/register/', payload)
            expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/create/', payload2)
            expect(wrapper.instance().state.redirect_profile).toBe(true)
            expect(axios.post).toHaveBeenCalledTimes(2)
        })

        it('should create teacher account and redirect to profile page', async()=>{
            let wrapper = shallow(<SignUpApp.WrappedComponent {...props2} />)
            wrapper.instance().validateData = jest.fn()
            wrapper.instance().validateData.mockReturnValueOnce(true)
            wrapper.instance().state.email = 'teacher@gmail.com'
            wrapper.instance().state.password = 'password'
            wrapper.instance().state.first_name = 'name'
            wrapper.instance().state.last_name = 'last'
            wrapper.instance().state.class_code = '123456'

            const payload = {
                username: 'teacher@gmail.com', 
                password: 'password', 
                first_name: 'name', 
                email: 'teacher@gmail.com'
            }

            const payload2 = {
                last_name:'last'
            }

            const user = mockResponseData('Token')
            const user2 = mockResponseData(true)
            
            axios.post.mockImplementationOnce(() => Promise.resolve(user))
            axios.post.mockImplementationOnce(() => Promise.resolve(user2))

            await wrapper.instance().handleSignup()
            
            expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//auth/register/', payload)
            expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//teachers/create/', payload2)
            expect(wrapper.instance().state.redirect_profile).toBe(true)
            expect(axios.post).toHaveBeenCalledTimes(2)
        })
    })
})

const mockResponseData = jest.fn((data) => {
    if (typeof data === 'string')
    {
        return {
            data: {token: data},
        };
    }
    else
    {
        return {
            data: data,
        };
    }
  });