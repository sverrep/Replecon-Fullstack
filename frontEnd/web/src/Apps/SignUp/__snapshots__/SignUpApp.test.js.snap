// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Up Tests Render Tests Student Render 1`] = `
Array [
  <Container
    className="signup-container"
    fluid={false}
  >
    <div
      className="vertical-center"
    >
      <Row>
        <h3
          className="signup-h3"
        >
          Student
           Sign Up
        </h3>
      </Row>
      <Row>
        <Row>
          <FormGroup
            className="mb-3"
            controlId="email"
          >
            <FloatingLabel
              label="Email"
            >
              <FormControl
                id="email"
                onChange={[Function]}
                placeholder="Email"
              />
            </FloatingLabel>
            <FloatingLabel
              label="Password"
            >
              <FormControl
                id="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
              />
            </FloatingLabel>
          </FormGroup>
        </Row>
        <Row>
          <FormGroup
            className="mb-3"
            controlId="email"
          >
            <FloatingLabel
              label="Name"
            >
              <FormControl
                id="name"
                onChange={[Function]}
                placeholder="Name"
              />
            </FloatingLabel>
            <FloatingLabel
              label="Class Code"
            >
              <FormControl
                id="class_code"
                onChange={[Function]}
                placeholder="Class code"
              />
            </FloatingLabel>
          </FormGroup>
        </Row>
      </Row>
      <p
        className="error-message"
      >
        
      </p>
      <div
        className="signup-btns-row"
      >
        <Button
          active={false}
          className="signup-btns"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          SignUp
        </Button>
      </div>
      <div
        className="signup-btns-row"
      >
        <Button
          active={false}
          className="signup-btns"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Already have an account? Log In
        </Button>
      </div>
    </div>
  </Container>,
]
`;

exports[`Sign Up Tests Render Tests Teacher Render 1`] = `
Array [
  <Container
    className="signup-container"
    fluid={false}
  >
    <div
      className="vertical-center"
    >
      <Row>
        <h3
          className="signup-h3"
        >
          Teacher
           Sign Up
        </h3>
      </Row>
      <Row>
        <Row>
          <FormGroup
            className="mb-3"
            controlId="email"
          >
            <FloatingLabel
              label="Email"
            >
              <FormControl
                id="email"
                onChange={[Function]}
                placeholder="Email"
              />
            </FloatingLabel>
            <FloatingLabel
              label="Password"
            >
              <FormControl
                id="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
              />
            </FloatingLabel>
          </FormGroup>
        </Row>
        <Row>
          <FormGroup
            className="mb-3"
            controlId="email"
          >
            <FloatingLabel
              label="First Name"
            >
              <FormControl
                id="first_name"
                onChange={[Function]}
                placeholder="First Name"
              />
            </FloatingLabel>
            <FloatingLabel
              label="Last Name"
            >
              <FormControl
                id="last_name"
                onChange={[Function]}
                placeholder="Last Name"
              />
            </FloatingLabel>
          </FormGroup>
        </Row>
      </Row>
      <p
        className="error-message"
      >
        
      </p>
      <div
        className="signup-btns-row"
      >
        <Button
          active={false}
          className="signup-btns"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          SignUp
        </Button>
      </div>
      <div
        className="signup-btns-row"
      >
        <Button
          active={false}
          className="signup-btns"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Already have an account? Log In
        </Button>
      </div>
    </div>
  </Container>,
]
`;

exports[`Sign Up Tests User Fields Test Common Field  1`] = `
<Row>
  <FormGroup
    className="mb-3"
    controlId="email"
  >
    <FloatingLabel
      label="Email"
    >
      <FormControl
        id="email"
        onChange={[Function]}
        placeholder="Email"
      />
    </FloatingLabel>
    <FloatingLabel
      label="Password"
    >
      <FormControl
        id="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </FloatingLabel>
  </FormGroup>
</Row>
`;

exports[`Sign Up Tests User Fields Test Student Unique Field  1`] = `
<Row>
  <FormGroup
    className="mb-3"
    controlId="email"
  >
    <FloatingLabel
      label="Name"
    >
      <FormControl
        id="name"
        onChange={[Function]}
        placeholder="Name"
      />
    </FloatingLabel>
    <FloatingLabel
      label="Class Code"
    >
      <FormControl
        id="class_code"
        onChange={[Function]}
        placeholder="Class code"
      />
    </FloatingLabel>
  </FormGroup>
</Row>
`;

exports[`Sign Up Tests User Fields Test Teacher Unique Field  1`] = `
<Row>
  <FormGroup
    className="mb-3"
    controlId="email"
  >
    <FloatingLabel
      label="First Name"
    >
      <FormControl
        id="first_name"
        onChange={[Function]}
        placeholder="First Name"
      />
    </FloatingLabel>
    <FloatingLabel
      label="Last Name"
    >
      <FormControl
        id="last_name"
        onChange={[Function]}
        placeholder="Last Name"
      />
    </FloatingLabel>
  </FormGroup>
</Row>
`;
